File: library-producer/src/main/java/com/ssscl/java/kafka/KafkaApplication.java
package com.ssscl.java.kafka;


import com.ssscl.java.kafka.messaging.MessageProducer;
import com.ssscl.kafka.common.Book;
import com.ssscl.kafka.common.Library;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.ConfigurableApplicationContext;
import org.springframework.context.annotation.Bean;
import org.springframework.kafka.core.KafkaTemplate;

import javax.annotation.Resource;

@SpringBootApplication
public class KafkaApplication {

    @Value(value = "${book.topic.name}")
    private String bookTopicName;

    @Value(value = "${library.topic.name}")
    private String libraryTopicName;

    @Autowired
    private KafkaTemplate<String, Book> bookKafkaTemplate;

    @Autowired
    private KafkaTemplate<String, Library> libraryKafkaTemplate;

    public static void main(String[] args) throws Exception {

        ConfigurableApplicationContext context = SpringApplication.run(KafkaApplication.class, args);

        MessageProducer<Book> bookMessageProducer = context.getBean(MessageProducer.class);
        bookMessageProducer.sendMessage(new Book("MyBook"));

//        MessageProducer<Library> libraryMessageProducer = context.getBean(MessageProducer.class);
//        libraryMessageProducer.sendMessage(new Library("MyLibrary"));

        context.close();
    }

    @Bean(name="bookMessageProducer")
    public MessageProducer<Book> bookProducer() {
        bookKafkaTemplate.setDefaultTopic(bookTopicName);
        return new MessageProducer<Book>(bookKafkaTemplate);
    }

//    @Bean(name="libraryMessageProducer")
//    public MessageProducer<Library> libraryProducer() {
//        libraryKafkaTemplate.setDefaultTopic(libraryTopicName);
//        return new MessageProducer<Library>(libraryKafkaTemplate);
//    }
}

File: library-consumer/src/main/java/com/ssscl/kotlin/kafka/KafkaApplication.kt
package com.ssscl.kotlin.kafka

import com.ssscl.kafka.common.Book
import com.ssscl.kafka.common.Library
import com.ssscl.kotlin.kafka.messaging.MessageConsumer
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.kafka.core.KafkaTemplate
import java.util.concurrent.TimeUnit

@SpringBootApplication
object KafkaApplication {

    @Value(value = "\${book.topic.name}")
    private val bookTopicName: String? = null

    @Value(value = "\${library.topic.name}")
    private val libraryTopicName: String? = null

    @Autowired
    private val bookKafkaTemplate: KafkaTemplate<String, Book>? = null

    @Autowired
    private val libraryKafkaTemplate: KafkaTemplate<String, Library>? = null

    @Throws(Exception::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val context = SpringApplication.run(KafkaApplication::class.java, *args)
        val messageConsumer: MessageConsumer = context.getBean(MessageConsumer::class.java)
        messageConsumer.bookLatch.await(10, TimeUnit.SECONDS);
        messageConsumer.libraryLatch.await(10, TimeUnit.SECONDS);
        context.close()
    }
}